app/src/main/java/org/mozilla/focus/MainActivity.kt

```diff
@@ -149,13 +151,36 @@ class MainActivity : LocaleAwareAppCompatActivity(), OnUrlEnteredListener, Media
     override fun onBackPressed() {
         val fragmentManager = supportFragmentManager
         val browserFragment = fragmentManager.findFragmentByTag(BrowserFragment.FRAGMENT_TAG) as BrowserFragment?
-        if (browserFragment != null &&
-                browserFragment.isVisible &&
-                browserFragment.onBackPressed()) {
-            // The Browser fragment handles back presses on its own because it might just go back
-            // in the browsing history.
-            return
+
+        if (browserFragment != null) {
+            if (browserFragment.isVisible &&
+                    browserFragment.onBackPressed()) {
+                // The Browser fragment handles back presses on its own because it might just go back
+                // in the browsing history.
+                return
+            }
+
+            val currFragment = fragmentManager.findFragmentByTag(SettingsFragment.FRAGMENT_TAG)
+
+            if (browserFragment.arguments == null) {
+                browserFragment.arguments = Bundle()
+            }
+
+            // Set ParentFragment flag to the BrowserFragment based on currFragment and let
+            // fragment lifecycle handle the rest
+            when (currFragment) {
+                is SettingsFragment -> {
+                    browserFragment.arguments!!.putSerializable("PARENT_FRAGMENT", BrowserNavigationOverlay.ParentFragment.SETTINGS)
```

Comment:
Nit: can you reuse this string as a constant? Just make it a companion object so there's less chance of typos causing other bugs.

Reply:
Updated @liuche